### PR TITLE
CloudWatch Logs Addon with Fluent Bit

### PR DESCRIPTION
**Pull request by elamaran11**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: lib/addons/cloudwatch-logs/index.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+ * @param helmOptions User provided values to customize the chart
+ */
+function populateValues(clusterInfo: ClusterInfo, helmOptions: CloudWatchLogsAddonProps): Values {
+    const values = helmOptions.values ?? {};
+    setPath(values, "serviceAccount.name", helmOptions.serviceAccountName);
+    setPath(values, "serviceAccount.create", false);
+    setPath(values, "cloudWatch.enabled", false);
+    setPath(values, "cloudWatchLogs.enabled", true);
+    setPath(values, "cloudWatchLogs.region", clusterInfo.cluster.stack.region);
+    setPath(values, "cloudWatchLogs.logGroupName", `/aws/eks/${clusterInfo.cluster.clusterName}/workloads`);
```
Comment (shapirov103) [inline lib/addons/cloudwatch-logs/index.ts:104]: It seems that customers may want to control the log group name, e.g. an enterprise has a clear convention on those that they would want to adhere. I think the default makes sense, but we should let them specify their own as well.

File: docs/addons/index.md [hunk 2]
```diff
@@ -22,6 +22,7 @@ The framework currently supports the following add-ons.
 | [`AWS Private CA Issuer`](./aws-privateca-issuer)                      | Installs AWS Private CA Issuer into your cluster.                                                                        |
 | [`CertManagerAddOn`](./cert-manager)                                   | Adds Certificate Manager to your EKS cluster.                                                             |
 | [`CalicoOperatorAddOn`](./calico-operator)                             | Adds the Calico CNI/Network policy cluster.                                                                                       |
+| [`CloudWatchLogsAddOn`](./cloudwatch-logs.md)                       | Adds Fluent log processor and forwarder to forward dataplane and application logs to Amazon CloudWatch.                                          |
```
Comment (shapirov103) [inline docs/addons/index.md:25]: did you mean Fluentbit log processor?

File: lib/addons/cloudwatch-logs/index.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import { getCloudWatchLogsPolicyDocument } from "./iam-policy";
+
+/**
+ * Configuration options for the FluentBit add-on.
+ */
+export interface CloudWatchLogsAddonProps extends HelmAddOnUserProps {
+    /**
+     * Iam policies for the add-on.
+     */
+    policyDocumentProvider?: (partition: string) => PolicyDocument;
```
Comment (shapirov103) [inline lib/addons/cloudwatch-logs/index.ts:17]: Why is this here? Can you describe a case when customer will pass some additional policies here?
Also I don't see that you honor any input in this field from the customers. Regardless of what they provide you will apply your own.

File: lib/addons/cloudwatch-logs/index.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    readonly options: CloudWatchLogsAddonProps;
+
+    constructor(props?: CloudWatchLogsAddonProps) {
+        super({ ...defaultProps as any, ...props });
+        this.options = this.props;
+    }
+
+    deploy(clusterInfo: ClusterInfo): Promise<Construct> {
+        let values: Values = populateValues(clusterInfo, this.options);
```
Comment (shapirov103) [inline lib/addons/cloudwatch-logs/index.ts:65]: I don't see merging values that customers may have passed. If customers pass values, we should honor them and only apply our defaults for they keys they did not specify.

File: lib/addons/cloudwatch-logs/index.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    logRetentionDays: 90,
+    values: {}
+};
+
+/**
+ * CloudWatchLogsAddon deploys FluentBit into an EKS cluster using the `aws-for-fluent-bit` Helm chart.
+ * https://github.com/aws/eks-charts/tree/master/stable/aws-for-fluent-bit
+ * 
+ */
+export class CloudWatchLogsAddon extends HelmAddOn {
```
Comment (shapirov103) [inline lib/addons/cloudwatch-logs/index.ts:55]: Is this add-on mutually exclusive with aws-fluentbit-addon? if so, we should add a check and fail fast.

File: lib/addons/cloudwatch-logs/index.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import { setPath } from "../../utils";
+import { HelmAddOn, HelmAddOnUserProps } from "../helm-addon";
+import { ClusterInfo, Values } from "../../spi/types";
+import { createNamespace } from "../../utils/namespace-utils";
+import { getCloudWatchLogsPolicyDocument } from "./iam-policy";
+
+/**
+ * Configuration options for the FluentBit add-on.
+ */
+export interface CloudWatchLogsAddonProps extends HelmAddOnUserProps {
```
Comment (shapirov103) [inline lib/addons/cloudwatch-logs/index.ts:13]: Are there any performance settings that we need to expose? I recall this particular component becoming a bottleneck in large clusters. 
We don't necessarily have to expose all the performance controls. At the very least we should honor any values that customers provide. If we can add a paragraph to the doc about performance tuning that would be awesome (e.g. with a reference to public docs on fluentbit performance tuning).

**Review by shapirov103 — COMMENTED** [review]
Awesome bias for action! Please see a few comments.

File: lib/addons/cloudwatch-logs/index.ts (hunk 1 above)
Comment (elamaran11) [reply lib/addons/cloudwatch-logs/index.ts]: Fixed.

File: docs/addons/index.md (hunk 2 above)
Comment (elamaran11) [reply docs/addons/index.md]: Fixed.

File: lib/addons/cloudwatch-logs/index.ts (hunk 4 above)
Comment (elamaran11) [reply lib/addons/cloudwatch-logs/index.ts:67]: Fixed.

File: lib/addons/cloudwatch-logs/index.ts (hunk 6 above)
Comment (shapirov103) [reply lib/addons/cloudwatch-logs/index.ts:13]: ?

Comment (elamaran11) [reply lib/addons/cloudwatch-logs/index.ts:13]: @shapirov103 I did add a section to doc page on how to do performance tuning. We dont want to support lot of parameters for this but the recommendation will help customers to pass values to achieve it. Please check

File: lib/addons/cloudwatch-logs/index.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+const defaultProps: CloudWatchLogsAddonProps = {
+    name: 'aws-for-fluent-bit',
+    chart: 'aws-for-fluent-bit',
+    release: "blueprints-addon-aws-fluent-bit-for-cw",
+    version: '0.1.24',
+    repository: 'https://aws.github.io/eks-charts',
+    namespace: 'aws-for-fluent-bit',
+    createNamespace: true,
+    serviceAccountName: 'aws-fluent-bit-for-cw-sa',
+    logGroupPrefix: '/aws/eks/blueprints-construct-dev',
```
Comment (shapirov103) [inline lib/addons/cloudwatch-logs/index.ts:46]: Why hardcoded prefix?

Comment (elamaran11) [reply lib/addons/cloudwatch-logs/index.ts:46]: What else is your proposal? You wanted it be passed by the user right?

Comment (shapirov103) [reply lib/addons/cloudwatch-logs/index.ts:46]: Yes,  I cannot imagine any customer who will ever want to see a log group prefx that states blueprint-construct-dev. 
I can however, imagine a prefix that is defaulted to the `eks/${clusterName}`. So let's do that. 
You can set the value of the default as "auto" and set it to the cluster name. Or customers can pass it directly.

File: docs/addons/cloudwatch-logs.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
+# Amazon CloudWatch Logs Addon using Fluent Bit.
+
+Fluent Bit is an open source Log Processor and Forwarder which allows you to collect any data like metrics and logs from different sources, enrich them with filters and send them to multiple destinations.
+
+AWS provides a Fluent Bit image with plugins for CloudWatch Logs, Kinesis Data Firehose, Kinesis Data Stream and Amazon OpenSearch Service.
+
+This add-on is configured to stream the worker node logs to CloudWatch Logs. The [AWS for Fluent Bit image](https://gallery.ecr.aws/aws-observability/aws-for-fluent-bit) is available on the Amazon ECR Public Gallery. For more details, see [AWS for Fluent Bit GitHub repository](https://github.com/aws/aws-for-fluent-bit).
```
Comment (shapirov103) [inline docs/addons/cloudwatch-logs.md:7]: Let's call out that this approach will handle node and application level logging. App logging will automatically be handled for containers emitting to std out /err.

File: docs/addons/cloudwatch-logs.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+Navigate to Log groups in Amazon CloudWatch console to see log groups starting with `/aws/eks/<your-cluster-name>/workload`.
+
+## Functionality
+
+Applies the CloudWatch Logs add-on to an Amazon EKS cluster. 
+
+## Performance Tuning
+
+By default, we send Fluent Bit application logs and Kubernetes metadata to CloudWatch. If you want to reduce the volume of data being sent to CloudWatch, you can stop one or both of these data sources from being sent to CloudWatch. Please take a look at Amazon CloudWatch documentation on [Reducing the log volume from Fluent Bit](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Container-Insights-setup-logs-FluentBit.html) and please look at filter section of aws-for-fluent-bit [chart](https://github.com/aws/eks-charts/blob/master/stable/aws-for-fluent-bit/values.yaml) to find ways to apply filters to performance tune your logs that is sent to Cloudwatch.
```
Comment (shapirov103) [inline docs/addons/cloudwatch-logs.md:57]: minor: 
`and please look at THE filter section`
`to performance tune your logs that ARE sent`

**Review by shapirov103 — COMMENTED** [review]
Almost there, please see my comments.

File: docs/addons/cloudwatch-logs.md (hunk 9 above)
Comment (elamaran11) [reply docs/addons/cloudwatch-logs.md]: Fixed

**Comment by elamaran11** [comment]
> Almost there, please see my comments.

@shapirov103 All issues fixed. Please review and run `e2e`.

File: docs/addons/cloudwatch-logs.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
+# Amazon CloudWatch Logs Addon using Fluent Bit.
+
+Fluent Bit is an open source Log Processor and Forwarder which allows you to collect any data like metrics and logs from different sources, enrich them with filters and send them to multiple destinations.
+
+AWS provides a Fluent Bit image with plugins for CloudWatch Logs, Kinesis Data Firehose, Kinesis Data Stream and Amazon OpenSearch Service.
+
+This add-on will take care of handling both node and application level logging. THis add-on is configured to stream the worker node logs to CloudWatch Logs. The [AWS for Fluent Bit image](https://gallery.ecr.aws/aws-observability/aws-for-fluent-bit) is available on the Amazon ECR Public Gallery. For more details, see [AWS for Fluent Bit GitHub repository](https://github.com/aws/aws-for-fluent-bit).
```
Comment (shapirov103) [inline docs/addons/cloudwatch-logs.md:7]: Minor: THis = This
How about changing "THis add-on" to "It"

**Review by shapirov103 — COMMENTED** [review]
Minor docs comments.

File: docs/addons/cloudwatch-logs.md [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+```
+NAME                                                                  READY   STATUS    RESTARTS   AGE
+pod/blueprints-addon-aws-fluent-bit-for-cw-aws-for-fluent-bit-46kkk   1/1     Running   0          3m15s
+pod/blueprints-addon-aws-fluent-bit-for-cw-aws-for-fluent-bit-6x4xq   1/1     Running   0          3m18s
+
+NAME                                                                       DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
+daemonset.apps/blueprints-addon-aws-fluent-bit-for-cw-aws-for-fluent-bit   2         2         2       2            2           <none>          3h53m
+```
+
+Navigate to Log groups in Amazon CloudWatch console to see log groups starting with `/aws/eks/<your-cluster-name>/workload`.
```
[the remark this answers is not in the document]
Comment (elamaran11) [reply docs/addons/cloudwatch-logs.md:50]: This is still correct because till cluster name is HLQ.

**Comment by elamaran11** [comment]
@shapirov103 I see the stack completed and deleted fine but we still dont see a comment on the PR.

**Review by shapirov103 — APPROVED** [review]
LGTM
